### ISSUE TITLE
Error when building newest polytracker version from source

### ISSUE THREAD

**Issue by BwRy** [trailofbits/polytracker#53]
OS: Ubuntu 18
Compiler: clang 7.1.0
Polytracker version: newest via git pull
Issue: When compiling the newest version of polytracker, it throws several errors with similar statements, e.g.
"error: no type named 'mutex' in namespace 'std'"

Full error:
~/Desktop/tools/taint/polytracker/build$ cmake -G Ninja -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ .. && ninja install
-- LLVM_VERSION_MAJOR: 7
-- LLVM_VERSION_MINOR: 1
-- LLVM_VERSION_PATCH: 0
-- ====ADDING sanitizer_linux_x86====
-- LIBBAMES ARE dfsan_rt-x86_64
-- Adding library dfsan_rt-x86_64
-- Adding extra compile flags: -stdlib=libc++;-nostdinc++ -g;-fPIC
-- Adding extra link flags: -Wl,--whole-archive /polytracker/polytreacker/dfsan/dfsan_rt/dfsan_private_headers/lib/libc++.a;/polytracker/polytreacker/dfsan/dfsan_rt/dfsan_private_headers/lib/libc++.a -Wl,--no-whole-archive
-- GOING FOR THE LINK
-- Looking for python3.7
-- Looking for python3.7 - found: /usr/local/bin/python3.7
-- Looking for python3.7 venv module
-- Looking for python3.7 venv module - found
-- Configuring done
-- Generating done
-- Build files have been written to: ~/Desktop/tools/taint/polytracker/build
[2/9] Building CXX object polytracker/dfsan/dfsan_sources/CMakeFiles/TaintSources.dir/taint_management.cpp.o
FAILED: polytracker/dfsan/dfsan_sources/CMakeFiles/TaintSources.dir/taint_management.cpp.o
/usr/local/bin/clang++   -I../polytracker/dfsan -I../polytracker/includes -I../polytracker/dfsan/dfsan_includes -I../polytracker/dfsan/dfsan_rt -I/polytracker/polytracker/dfsan/dfsan_rt/dfsan_private_headers/include/c++/v1 -g -stdlib=libc++ -nostdinc++ -fPIC -MD -MT polytracker/dfsan/dfsan_sources/CMakeFiles/TaintSources.dir/taint_management.cpp.o -MF polytracker/dfsan/dfsan_sources/CMakeFiles/TaintSources.dir/taint_management.cpp.o.d -o polytracker/dfsan/dfsan_sources/CMakeFiles/TaintSources.dir/taint_management.cpp.o -c ../polytracker/dfsan/dfsan_sources/taint_management.cpp
clang-7: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
In file included from ../polytracker/dfsan/dfsan_sources/taint_management.cpp:4:
../polytracker/dfsan/dfsan_includes/taint_management.hpp:52:8: error: no type named 'mutex' in namespace 'std'
                std::mutex taint_info_mutex;
                ~~~~~^
1 error generated.

**Comment by notBD** [comment]
Hi, have you tried building from within the docker image? Apologies for the outdated readme, but we currently only build in docker. 

Could you try running from the main directory `docker build -t trailofbits/polytracker .` and let me know if it builds for you? 

Thanks

**Comment by notBD** [comment]
If you would like to try running polytracker outside of docker, I recommend that temporarily you build polytracker inside of docker.
 
1. Then you can copy the install(`/polytracker/build/bin/polytracker`) to your machine. 
2. Use polyclang to compile your target 

Then set POLYPATH and I think things should just work. If that doesnt work for you, please let me know. 

When we have time we will try and fix Polytracker so you can build from source easily again.

**Comment by BwRy** [comment]
I get a different error message inside the docker container when attempting to run polyclang:

root@5dc41f4a991b:/polytracker# $CC
clang: warning: argument unused during compilation: '-mllvm -polytrack-dfsan-abilist=/polytracker/build/bin/polytracker//abi_lists/polytrack_abilist.txt' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-mllvm -polytrack-dfsan-abilist=/polytracker/build/bin/polytracker//abi_lists/dfsan_abilist.txt' [-Wunused-command-line-argument]
whi/usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../x86_64-linux-gnu/Scrt1.o: In function `_start':
(.text+0x20): undefined reference to `main'
chclang: error: linker command failed with exit code 1 (use -v to see invocation)
root@5dc41f4a991b:/polytracker# which $CC
/polytracker/build/bin/polytracker/polyclang
root@5dc41f4a991b:/polytracker#

**Comment by BwRy** [comment]
during the docker build command:

[168/180] Linking CXX executable polytracker/polyclang/polyclang
[169/180] Building CXX object polytracker/dfsan/dfsan_sources/CMakeFiles/TaintSources.dir/taint_sources.cpp.o
clang: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
[170/180] Building CXX object polytracker/dfsan/dfsan_sources/CMakeFiles/TaintSources.dir/taint_management.cpp.o
clang: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
[171/180] Linking CXX static library polytracker/dfsan/dfsan_sources/libTaintSources.a
[172/180] Building CXX object polytracker/dfsan/dfsan_rt/dfsan/CMakeFiles/dfsan_rt-x86_64.dir/dfsan_interceptors.cc.o
clang: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
[173/180] Building CXX object polytracker/dfsan/dfsan_rt/dfsan/CMakeFiles/dfsan_rt-x86_64.dir/dfsan_custom.cc.o
clang: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
[174/180] Building CXX object polytracker/dfsan/dfsan_rt/dfsan/CMakeFiles/dfsan_rt-x86_64.dir/dfsan.cc.o
clang: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
[175/180] Building CXX object polytracker/dfsan/dfsan_pass/CMakeFiles/DataFlowSanitizerPass.dir/DataFlowSanitizer.cpp.o
[176/180] Linking CXX shared module polytracker/dfsan/dfsan_pass/libDataFlowSanitizerPass.so
[177/180] Building CXX object polytracker/dfsan/dfsan_rt/dfsan/CMakeFiles/dfsan_rt-x86_64.dir/dfsan_log_mgmt.cpp.o
clang: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
[178/180] Linking CXX static library polytracker/dfsan/dfsan_rt/libdfsan_rt-x86_64.a
[179/180] Generating exported symbols for dfsan_rt-x86_64

**Comment by BwRy** [comment]
In the docker image, I then removed the build directory and tried to manually build (within docker):
root@69a878fe9e71:/polytracker/# mkdir build && cd build
root@69a878fe9e71:/polytracker/build/# cmake -G Ninja -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ .. && ninja install
<...snip...>
-- Build files have been written to: /polytracker/build
[169/180] Building CXX object polytracker/dfsan/dfsan_sources/CMakeFiles/TaintSources.dir/taint_sources.cpp.o
clang: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
[170/180] Building CXX object polytracker/dfsan/dfsan_sources/CMakeFiles/TaintSources.dir/taint_management.cpp.o
clang: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
[172/180] Building CXX object polytracker/dfsan/dfsan_rt/dfsan/CMakeFiles/dfsan_rt-x86_64.dir/dfsan_interceptors.cc.o
clang: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
[173/180] Building CXX object polytracker/dfsan/dfsan_rt/dfsan/CMakeFiles/dfsan_rt-x86_64.dir/dfsan_custom.cc.o
clang: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
[174/180] Building CXX object polytracker/dfsan/dfsan_rt/dfsan/CMakeFiles/dfsan_rt-x86_64.dir/dfsan.cc.o
clang: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
[177/180] Building CXX object polytracker/dfsan/dfsan_rt/dfsan/CMakeFiles/dfsan_rt-x86_64.dir/dfsan_log_mgmt.cpp.o
clang: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
[179/180] Install the project...
-- Install configuration: ""
-- Installing: /polytracker/build/bin/polytracker/polyclang
-- Installing: /polytracker/build/bin/polytracker/polyclang++
-- Installing: /polytracker/build/bin/polytracker/abi_lists/polytrack_abilist.txt
-- Installing: /polytracker/build/bin/polytracker/pass/libDataFlowSanitizerPass.so
-- Installing: /polytracker/build/bin/polytracker/lib/libdfsan_rt-x86_64.a
-- Installing: /polytracker/build/bin/polytracker/lib/libdfsan_rt-x86_64.a.syms
-- Up-to-date: /polytracker/build/bin/polytracker/lib/libdfsan_rt-x86_64.a.syms
-- Installing: /polytracker/build/bin/polytracker/abi_lists/dfsan_abilist.txt
-- Installing: /polytracker/build/bin/polytracker/lib/libTaintSources.a
root@69a878fe9e71:/polytracker/build# which polyclang
root@69a878fe9e71:/polytracker/build# polyclang
bash: polyclang: command not found
root@69a878fe9e71:/polytracker/build#cd bin/polytracker
root@69a878fe9e71:/polytracker/build/bin/polytracker# ./polyclang
clang: warning: argument unused during compilation: '-mllvm -polytrack-dfsan-abilist=/polytracker/build/bin/polytracker//abi_lists/polytrack_abilist.txt' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-mllvm -polytrack-dfsan-abilist=/polytracker/build/bin/polytracker//abi_lists/dfsan_abilist.txt' [-Wunused-command-line-argument]
/usr/lib/gcc/x86_64-linux-gnu/7.4.0/../../../x86_64-linux-gnu/Scrt1.o: In function `_start':
(.text+0x20): undefined reference to `main'
clang: error: linker command failed with exit code 1 (use -v to see invocation)

**Comment by notBD** [comment]
So for the first error, when you are just invoking $CC it is expecting you to give it a target file to compile. It cannot find one so it exits to main. 

Polyclang is just a wrapper around clang that has it load our LLVM pass and link some libraries. 

Try running `./polyclang -g -o some_file some_file.c` where some_file.c is a toy c program.

**Comment by BwRy** [comment]
Thanks! I got it working on a toy file. On an actual project, I had to edit the "ac_fn_c_try_compile" function in autoconf's configure file to get ./configure to work.

I should probably open a different issue now, as polyclang is also being used for linking/LD when compiling:

make[2]: Entering directory
 <...snip...>
clang: error: unsupported option '--whole-archive'
clang: error: unsupported option '--no-whole-archive'
clang: error: unknown argument: '-soname'
